File: backend/app/schemas/nearby_market.py
import uuid
from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class NearbyMarketRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    radius_km: int = Field(strict=True, ge=1, le=10)
    business_query: Optional[str] = Field(default=None, min_length=2, max_length=200)
    business_slug: Optional[str] = Field(default=None, min_length=1, max_length=100, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


    @field_validator('business_query', mode='before')
    @classmethod
    def clean_query(cls, value):
        return ' '.join(value.split()) if isinstance(value, str) else value

    @model_validator(mode='after')
    def require_intent(self):
        if not self.business_query and not self.business_slug:
            raise ValueError('Business query or catalog business is required')
        return self


class EvidenceWarning(BaseModel):
    code: str
    message: str


class ResolvedLocation(BaseModel):
    latitude: float
    longitude: float
    display_name: str
    query_used: str
    provider: str
    precision: str
    address: Dict[str, str]
    location_fingerprint: str
    fetched_at: datetime
    expires_at: datetime


class NearbyBusiness(BaseModel):
    id: Optional[uuid.UUID] = None
    slug: Optional[str] = None
    display_name: str


class NearbyPOI(BaseModel):
    provider: str
    external_type: Literal["node", "way", "relation", "place"]
    external_id: str
    name: str
    latitude: float
    longitude: float
    coordinate_kind: Literal["NODE", "REPRESENTATIVE_CENTER", "PLACE_LOCATION"]
    classification: Literal["DIRECT_COMPETITOR", "RELATED_BUSINESS", "GENERIC_POI"]
    matched_business_slug: str
    matching_rule: str
    matching_evidence: List[Dict[str, str]]
    normalized_tags: Dict[str, str]
    normalized_address: Dict[str, str]
    fetched_at: datetime
    distance_meters: float
    distance_km: str


class NearbySummary(BaseModel):
    direct_competitors: int
    related_businesses: int
    nearest_direct_competitor: Optional[NearbyPOI]


class NearbyRadius(BaseModel):
    selected_km: int
    selected_meters: int
    distance_method: Literal["POSTGIS_GEOGRAPHY_SPHEROID"] = "POSTGIS_GEOGRAPHY_SPHEROID"


class NearbySource(BaseModel):
    provider: Literal["OPENSTREETMAP_OVERPASS", "GOOGLE_PLACES"] = "OPENSTREETMAP_OVERPASS"
    mode: Literal["LIVE", "CACHE"]
    cache_status: Literal["FRESH_FETCH", "FRESH_CACHE", "STALE_CACHE"]
    fetched_at: datetime
    expires_at: datetime
    mapping_version: str


class NearbyQuality(BaseModel):
    complete_query: bool
    rejected_record_count: int
    warnings: List[EvidenceWarning]
    attribution: str = "© OpenStreetMap contributors (ODbL)"


class NearbyMarketResponse(BaseModel):
    business_query: Optional[str] = None
    matched_catalog_business_slug: Optional[str] = None
    business: NearbyBusiness
    location: ResolvedLocation
    radius: NearbyRadius
    summary: NearbySummary
    competitors: List[NearbyPOI]
    related_businesses: List[NearbyPOI]
    generic_pois: List[NearbyPOI] = Field(default_factory=list)
    source: NearbySource
    quality: NearbyQuality

File: src/types/nearbyMarket.ts
export interface NearbyPOI {
  provider: string
  external_type: 'node' | 'way' | 'relation' | 'place'
  external_id: string
  name: string
  latitude: number
  longitude: number
  coordinate_kind: 'NODE' | 'REPRESENTATIVE_CENTER' | 'PLACE_LOCATION'
  classification: 'DIRECT_COMPETITOR' | 'RELATED_BUSINESS' | 'GENERIC_POI'
  matched_business_slug: string
  matching_rule: string
  matching_evidence: Array<{ key: string; value: string }>
  normalized_tags: Record<string, string>
  normalized_address: Record<string, string>
  fetched_at: string
  distance_meters: number
  distance_km: string
}

/** Provider evidence keys are preserved; unlike legacy demo data, tags are not camelized. */
export interface NearbyMarketEvidence {
  business_query?: string
  matched_catalog_business_slug?: string | null
  generic_pois?: NearbyPOI[]
  business: { id: string | null; slug: string | null; display_name: string }
  location: {
    latitude: number; longitude: number; display_name: string; query_used: string
    provider: string; precision: string; location_fingerprint: string
    fetched_at: string; expires_at: string; address: Record<string, string>
  }
  radius: { selected_km: number; selected_meters: number; distance_method: 'POSTGIS_GEOGRAPHY_SPHEROID' }
  summary: { direct_competitors: number; related_businesses: number; nearest_direct_competitor: NearbyPOI | null }
  competitors: NearbyPOI[]
  related_businesses: NearbyPOI[]
  source: { provider: string; mode: 'LIVE' | 'CACHE'; cache_status: 'FRESH_FETCH' | 'FRESH_CACHE' | 'STALE_CACHE'; fetched_at: string; expires_at: string; mapping_version: string }
  quality: { complete_query: boolean; rejected_record_count: number; warnings: Array<{ code: string; message: string }>; attribution: string }
}
